controller/sync.go

```diff
@@ -441,184 +492,160 @@ func hasCRDOfGroupKind(resources []managedResource, group string, kind string) b
 	return false
 }
 
-// performs a apply based sync of the given sync tasks (possibly pruning the objects).
-// If update is true, will updates the resource details with the result.
-// Or if the prune/apply failed, will also update the result.
-func (sc *syncContext) doApplySync(syncTasks []syncTask, dryRun, force, update bool) bool {
-	syncSuccessful := true
-
-	var createTasks []syncTask
-	var pruneTasks []syncTask
-	for _, syncTask := range syncTasks {
-		if syncTask.targetObj == nil {
-			pruneTasks = append(pruneTasks, syncTask)
+// terminate looks for any running jobs/workflow hooks and deletes the resource
+func (sc *syncContext) terminate() {
+	terminateSuccessful := true
+	sc.log.Debug("terminating")
+	tasks, _ := sc.getSyncTasks()
+	for _, task := range tasks {
+		if !task.isHook() || !task.completed() {
+			continue
+		}
+		if isRunnable(task.groupVersionKind()) {
+			err := sc.deleteResource(task)
+			if err != nil {
+				sc.setResourceResult(task, "", OperationFailed, fmt.Sprintf("Failed to delete: %v", err))
+				terminateSuccessful = false
+			} else {
+				sc.setResourceResult(task, "", OperationSucceeded, fmt.Sprintf("Deleted"))
+			}
+		}
+	}
+	if terminateSuccessful {
+		sc.setOperationPhase(OperationFailed, "Operation terminated")
+	} else {
+		sc.setOperationPhase(OperationError, "Operation termination had errors")
+	}
+}
+
+func (sc *syncContext) deleteResource(task *syncTask) error {
+	sc.log.WithFields(log.Fields{"task": task}).Debug("deleting task")
+	apiResource, err := kube.ServerResourceForGroupVersionKind(sc.disco, task.groupVersionKind())
+	if err != nil {
+		return err
+	}
+	resource := kube.ToGroupVersionResource(task.groupVersionKind().GroupVersion().String(), apiResource)
+	resIf := kube.ToResourceInterface(sc.dynamicIf, apiResource, resource, task.namespace())
+	propagationPolicy := metav1.DeletePropagationForeground
+	return resIf.Delete(task.name(), &metav1.DeleteOptions{PropagationPolicy: &propagationPolicy})
+}
+
+var operationPhases = map[ResultCode]OperationPhase{
+	ResultCodeSynced:       OperationRunning,
+	ResultCodeSyncFailed:   OperationFailed,
+	ResultCodePruned:       OperationSucceeded,
+	ResultCodePruneSkipped: OperationSucceeded,
+}
+
+func (sc *syncContext) runTasks(tasks syncTasks, dryRun bool) (successful bool) {
```

Comment:
Style: both @alexmt and I discourage named return values unless really necessary (e.g. used in a deferred function).